kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java

```diff
@@ -0,0 +1,138 @@
+     */
+    public static @NonNull DestroyableRawSecretKey byOwnershipTransfer(@NonNull String algorithm, @NonNull byte[] bytes) {
+        return new DestroyableRawSecretKey(algorithm, bytes);
+    }
+
+    /**
+     * Create a new key by creating a copy of the given key material.
+     * @param algorithm The key algorithm
+     * @param bytes The key material
+     * @return The new key
+     */
+    public static @NonNull DestroyableRawSecretKey byClone(@NonNull String algorithm, @NonNull byte[] bytes) {
+        return new DestroyableRawSecretKey(algorithm, Objects.requireNonNull(bytes).clone());
+    }
+
+    /**
+     * Convert the given key to a destroyable key, attempting to destroy the given key.
+     * @param source The key to convert
+     * @return The new destroyable key.
+     */
+    public static @NonNull DestroyableRawSecretKey toDestroyableKey(@NonNull SecretKey source) {
+        Objects.requireNonNull(source);
+        if (!"RAW".equalsIgnoreCase(source.getFormat())) {
+            throw new IllegalArgumentException("RAW-format key required");
+        }
+        // no need to copy, because getEncoded should itself return a copy of the key material
+        var result = byOwnershipTransfer(source.getAlgorithm(), source.getEncoded());
+        try {
+            source.destroy();
+        }
+        catch (DestroyFailedException e) {
+            LOGGER.warn("Failed to destroy key of {}", source.getClass());
+        }
+        return result;
+    }
+
+    @Override
+    public @NonNull String getAlgorithm() {
+        return algorithm;
+    }
+
+    @Override
+    public @NonNull String getFormat() {
+        return "RAW";
+    }
+
+    /**
+     * Returns the RAW-encoded key.
+     * This is a copy the key. It is the callers responsibility to destroy this key material
+     * when it's no longer needed.
+     * @return The RAW-encoded key.
+     * @throws IllegalStateException if the key has been destroyed
+     */
+    @Override
+    public @NonNull byte[] getEncoded() {
+        checkNotDestroyed();
+        return key.clone();
+    }
+
+    private void checkNotDestroyed() {
+        if (isDestroyed()) {
+            throw new IllegalStateException();
+        }
+    }
+
+    @Override
+    public void destroy() {
+        Arrays.fill(key, (byte) 0);
+        destroyed = true;
+    }
+
+    @Override
+    public boolean isDestroyed() {
+        return destroyed;
+    }
+
+    /**
+     * Tests whether the arguments represent the same key.
+     * @param thisKey The one key
+     * @param thatKey The other key
+     * @return true if they keys have the same algorithm and key material.
+     */
+    public static boolean same(@NonNull DestroyableRawSecretKey thisKey, @NonNull DestroyableRawSecretKey thatKey) {
+        if (thisKey == thatKey) {
+            return true;
+        }
+        Objects.requireNonNull(thisKey).checkNotDestroyed();
+        Objects.requireNonNull(thatKey).checkNotDestroyed();
```

Comment:
Feels a bit odd for these to throw rather than return false especially when this is being consumed in lambdas. Is there a reason not to include `isDestoryed` in the return condition?

Reply:
Anything that looks like an equality operation gets very slippery if the operands have some state which on which the equality depends but which can change. In this instance 

* should `same(foo, foo)` return true or false for a destroyed `foo`? Returning false is seems weird, but we can make it return true by using `==`. 
* But then what about `(foo, bar)` when `foo` and `bar` are distinct objects that _had_ the same bytes? We can't make that return true. 

However, your comment prompted me to move this to become a method of the test class, rather than adding a method to the production code which has these issues.